source/common/network/interval_tree.h

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+#include "source/common/common/assert.h"
+
+#include "absl/strings/str_cat.h"
+#include "absl/types/optional.h"
+#include "absl/types/span.h"
+
+namespace Envoy {
+namespace Network {
+namespace IntervalTree {
+
+template <class Data, class N> class IntervalTree {
```

Comment:
It's very difficult/impossible to read this kind of code and understand what is going on as a new reader. Can you please add a large block comment at the top that describes what this does, provides a high level implementation overview, etc.? Also in general I would over comment the code below if possible.

Also in general this kind of code scares me a bit so you might consider writing a small fuzz test for it to help look for corner cases.